### PR TITLE
fix: handle invariant interrupts and assertions

### LINKED ISSUE #15684: invariant testing: SIGINT ignored after long campaigns on some targets — killed process loses the end-of-run summary, silently dropping assertion failures
**Version:** `forge 1.7.2-nightly` (`907ba0817e8f`), Linux x86_64 (Ubuntu on EC2 and WSL2)

### Bug

We benchmark forge with a fixed wall-clock budget: `forge test --mc CryticToFoundry --show-progress`, then SIGINT at budget expiry to trigger the graceful early exit. On some targets, once the campaign has run ~30+ min, SIGINT (repeated ones too) is never acted on — the process fuzzes at full CPU forever and must be SIGKILLed. Same harness, 2h budgets: three targets exit gracefully, two wedge reliably (7/7 cloud instances + deterministic WSL2 repro). Reproduces with `FOUNDRY_INVARIANT_SHRINK_RUN_LIMIT=0`, so inline shrinking is ruled out.

A SIGKILLed run has no end-of-run summary in the log, which is the **only** place handler assertion failures and pre-flight-failing invariants are reported (mid-run JSON `failure` events cover only invariant violations found during the campaign) — so those findings are silently lost and the run looks clean.

Ending the campaign from the inside via `[invariant] timeout` works on the same targets: full summary printed. The wind-down is fine; the SIGINT path is the problem.

### Repro

```bash
git clone https://github.com/scfuzzbench-org/superform-v2-periphery-scfuzzbench
cd superform-v2-periphery-scfuzzbench && git checkout 4d56ecd
export FOUNDRY_DYNAMIC_TEST_LINKING=false FOUNDRY_INVARIANT_SHRINK_RUN_LIMIT=0
forge test --mc CryticToFoundry --show-progress &
sleep 2400 && kill -INT %1   # no effect, ever
```

### Asks

1. Fix the SIGINT/early-exit path for long invariant campaigns (happy to provide a gdb thread dump of a wedged process — the repro is deterministic).
2. Also emit handler assertion failures as mid-run `failure` events, so a killed process doesn't lose already-discovered findings.

### PR DESCRIPTION
**Pull request by stevencartavia**
Allows `SIGINT` and other early-exit requests to interrupt an invariant campaign while a transaction is still executing.

Also emits handler assertion failures as deduplicated mid-run events so discovered findings are preserved if the final summary is not written.

Fixes #15684.

### CODE REVIEW

File: crates/evm/evm/src/executors/invariant/mod.rs [hunk 1]
```diff
@@ -1073,6 +1099,7 @@ impl<'a, FEN: FoundryEvmNetwork> InvariantExecutor<'a, FEN> {
         // suite runner so parameterized `invariant_*` functions are rejected with a per-test
```
Comment (mattsse) [inline crates/evm/evm/src/executors/invariant/mod.rs:1105]: Please add a regression test that starts an invariant handler which remains in active EVM execution, triggers the shared `EarlyExit`, and asserts that the campaign terminates promptly. The current tests only cover handler-event serialization and deduplication, so the primary behavior fixed here could regress and leave `forge test` hanging indefinitely without any test failing.

**Review by mattsse — DISMISSED** [review]
The interrupt path needs regression coverage before this can merge.

File: crates/evm/evm/src/executors/invariant/mod.rs (hunk 1 above)
Comment (stevencartavia) [reply crates/evm/evm/src/executors/invariant/mod.rs:1105]: [b6a1600](https://github.com/foundry-rs/foundry/pull/15689/commits/b6a16000b9749308b685218a193c3ba58767a796)

File: crates/evm/evm/src/inspectors/stack.rs [hunk 2]
```diff
@@ -1055,6 +1065,11 @@ impl<FEN: FoundryEvmNetwork> InspectorStackRefMut<'_, FEN> {
         interpreter: &mut Interpreter,
         ecx: &mut FoundryContextFor<'_, FEN>,
     ) {
+        if self.inner.early_exit.as_ref().is_some_and(EarlyExit::should_stop) {
+            interpreter.halt(InstructionResult::Stop);
```
Comment (decofe) [inline crates/evm/evm/src/inspectors/stack.rs]: The executor-level active-EVM interruption is already covered by `b6a1600`; the remaining gap is only the invariant-campaign integration. Because that test manually calls `set_early_exit`, it would still pass if the campaign stopped wiring its shared `EarlyExit` into worker executors or stopped checking it after `execute_tx`. If we want that extra coverage without slowing CI, a one-worker campaign (`runs = 1`, `depth = 1`) with an infinite-loop handler and a test-only first-opcode channel/barrier would make the timing deterministic; trigger Ctrl-C once the first opcode is reached and require the campaign result within one second. On timeout, set the exit flag before joining so cleanup remains bounded.

Comment (grandizzy) [reply crates/evm/evm/src/inspectors/stack.rs]: test nit, we could live without it

**Review by grandizzy — APPROVED** [review]
makes sense

**Review by mablr — DISMISSED** [review]

File: crates/evm/evm/src/inspectors/stack.rs [hunk 3]
```diff
@@ -1055,6 +1065,11 @@ impl<FEN: FoundryEvmNetwork> InspectorStackRefMut<'_, FEN> {
         interpreter: &mut Interpreter,
         ecx: &mut FoundryContextFor<'_, FEN>,
     ) {
+        if self.inner.early_exit.as_ref().is_some_and(EarlyExit::should_stop) {
```
Comment (grandizzy) [inline crates/evm/evm/src/inspectors/stack.rs]: **[P2] Check cancellation before classifying invariant-call results**

`STOP` still runs `call_end` hooks, so an interrupted invariant predicate with a pending `vm.expectRevert()` or similar expectation can have its `Stop` rewritten into a revert. `assert_invariants` would then record a synthetic `BrokenInvariant`.

We can avoid threading `EarlyExit` through the invariant APIs by exposing a read-only query on the inspector and checking it before interpreting predicate/hook results. This also covers preflight because it uses `assert_invariants` after the flag has been installed.

```diff
diff --git a/crates/evm/evm/src/inspectors/stack.rs b/crates/evm/evm/src/inspectors/stack.rs
@@
     pub(crate) fn set_early_exit(&mut self, early_exit: EarlyExit) {
         self.early_exit = Some(early_exit);
     }

+    /// Returns whether cancellation has been requested for this execution.
+    #[inline]
+    pub(crate) fn early_exit_requested(&self) -> bool {
+        self.early_exit.as_ref().is_some_and(EarlyExit::should_stop)
+    }
+
     /// Sets the block for the relevant inspectors.

diff --git a/crates/evm/evm/src/executors/invariant/result.rs b/crates/evm/evm/src/executors/invariant/result.rs
@@
         let (call_result, success) = call_invariant_function(
             executor,
             invariant_contract.address,
             invariant.abi_encode_input(&[])?.into(),
         )?;
+        if executor.inspector().early_exit_requested() {
+            break;
+        }
         if !success {
@@
             let (inv_result, success) = call_invariant_function(
                 &invariant_run.executor,
                 invariant_contract.address,
                 invariant_contract.anchor().abi_encode_input(&[])?.into(),
             )?;
+            if invariant_run.executor.inspector().early_exit_requested() {
+                return Ok(ContinueOutcome { continues: true, broken: None });
+            }
             if success
@@
     let (call_result, success) =
         call_after_invariant_function(&invariant_run.executor, invariant_contract.address)?;
+    if invariant_run.executor.inspector().early_exit_requested() {
+        return Ok(None);
+    }
     // Fail the test case if `afterInvariant` doesn't succeed.
```

**Review by grandizzy — CHANGES_REQUESTED** [review]
Please check edge case below

File: crates/evm/evm/src/inspectors/stack.rs (hunk 3 above)
Comment (stevencartavia) [reply crates/evm/evm/src/inspectors/stack.rs]: [acfbc87](https://github.com/foundry-rs/foundry/pull/15689/commits/acfbc87a8725554e30301702bfa386f0d853bd91)

File: crates/evm/evm/src/executors/invariant/mod.rs [hunk 4]
```diff
@@ -1073,6 +1099,7 @@ impl<'a, FEN: FoundryEvmNetwork> InvariantExecutor<'a, FEN> {
...
         // failure entry before any campaign runs.
         let config = invariant_worker_config(config, plan.worker_id, worker_count);
+        executor.inspector_mut().set_early_exit(campaign_state.early_exit().clone());
```
Comment (grandizzy) [inline crates/evm/evm/src/executors/invariant/mod.rs]: **[P1] Include campaign timeout and terminal stop in active-EVM cancellation**

This installs only the global `EarlyExit` token in the inspector, while `InvariantCampaignState::should_stop()` also includes the campaign timeout and `terminal_stop`. An active transaction with a very large gas limit therefore still cannot be interrupted by `[invariant] timeout`; similarly, when one worker requests a terminal stop, another worker can remain inside EVM execution while Rayon waits for it. This leaves the timeout case described in #15684 unresolved.

Please expose a campaign-level cancellation handle to the inspector whose stop condition covers the global early exit, terminal stop, and deadline, and use that same condition for the post-`execute_tx` check.

File: crates/evm/evm/src/executors/invariant/result.rs [hunk 5]
```diff
@@ -207,6 +207,9 @@ pub(crate) fn assert_invariants<'a, FEN: FoundryEvmNetwork>(
             invariant_contract.address,
             invariant.abi_encode_input(&[])?.into(),
         )?;
+        if executor.inspector().early_exit_requested() {
```
Comment (grandizzy) [inline crates/evm/evm/src/executors/invariant/result.rs]: **[P2] Propagate cancellation instead of returning a normal invariant result**

Breaking here suppresses the synthetic invariant failure, but makes cancellation indistinguishable from a successful check. The caller can consequently persist corpus data and increment the completed-run counters even though the predicate never completed. The optimization and `afterInvariant` paths have the same behavior.

Please propagate an explicit cancelled outcome, or re-check the campaign stop condition immediately after `can_continue` and `assert_after_invariant` and exit before corpus/end-of-run accounting.

**Review by grandizzy — CHANGES_REQUESTED** [review]
The SIGINT path is improved, but campaign-local stop conditions still cannot interrupt active EVM execution, and interrupted predicate/hook calls are accounted as completed runs.

File: crates/evm/evm/src/executors/invariant/mod.rs [hunk 6]
```diff
@@ -1154,6 +1181,9 @@ impl<'a, FEN: FoundryEvmNetwork> InvariantExecutor<'a, FEN> {
                     &mut current_run.executor,
                     current_run.inputs.last().expect("checked above"),
                 )?;
+                if campaign_state.early_exit().should_stop() {
```
Comment (mablr) [inline crates/evm/evm/src/executors/invariant/mod.rs]: This checks the campaign-global flag after `execute_tx`, so it cannot distinguish a call that was halted by the inspector from one that completed normally just before SIGINT arrived. In the latter case, a genuine handler assertion/revert is dropped before `can_continue` can record or emit it, which loses the finding this PR is intended to preserve.

File: crates/evm/evm/src/executors/invariant/mod.rs (hunk 4 above)
Comment (stevencartavia) [reply crates/evm/evm/src/executors/invariant/mod.rs]: [343ce4b](https://github.com/foundry-rs/foundry/pull/15689/commits/343ce4b8d6747181b836ce5941eab59d6656d3a5)

File: crates/evm/evm/src/executors/invariant/result.rs (hunk 5 above)
Comment (stevencartavia) [reply crates/evm/evm/src/executors/invariant/result.rs]: [343ce4b](https://github.com/foundry-rs/foundry/pull/15689/commits/343ce4b8d6747181b836ce5941eab59d6656d3a5)

File: crates/evm/evm/src/executors/invariant/mod.rs (hunk 6 above)
Comment (stevencartavia) [reply crates/evm/evm/src/executors/invariant/mod.rs]: [3bc6d82](https://github.com/foundry-rs/foundry/pull/15689/commits/3bc6d82b9e495863a04551fdb717c1c76b97152e)

File: crates/evm/evm/src/executors/invariant/mod.rs [hunk 7]
```diff
@@ -1154,6 +1209,10 @@ impl<'a, FEN: FoundryEvmNetwork> InvariantExecutor<'a, FEN> {
                     &mut current_run.executor,
                     current_run.inputs.last().expect("checked above"),
                 )?;
+                if call_result.execution_cancelled {
+                    invariant_test.test_data.failures.clone_from(&failures_checkpoint);
```
Comment (grandizzy) [inline crates/evm/evm/src/executors/invariant/mod.rs]: **[P1] Preserve completed findings when a later call is cancelled**

This unconditional checkpoint restore can erase a genuine predicate failure found earlier in the same run. For example, predicate A can fail after call 1, the multi-predicate campaign continues, and call 2 is interrupted; restoring the pre-run checkpoint then deletes predicate A.

Restore the checkpoint only if no new invariant or handler finding completed. Otherwise retain and publish those findings while still aborting the partial run.

File: crates/evm/evm/src/executors/invariant/mod.rs [hunk 8]
```diff
@@ -1356,25 +1414,26 @@ impl<'a, FEN: FoundryEvmNetwork> InvariantExecutor<'a, FEN> {
                         }
                     };
 
+                    // A predicate or optimization call can be interrupted after the handler
+                    // completed. Do not accept the partial run as a successful one.
+                    if predicate_cancelled {
+                        if completed_predicate_finding {
+                            current_run.save_last_run_inputs = true;
+                            stop_after_run = true;
```
Comment (grandizzy) [inline crates/evm/evm/src/executors/invariant/mod.rs]: **[P1] Do not accept an interrupted run just to preserve its finding**

`stop_after_run = true` preserves the finding, but falls through to corpus persistence, `end_run`, completed-call accounting, and `increment_runs`. Since a later predicate was interrupted, this reintroduces the partial-run accounting issue.

Cancellation with a completed finding needs a separate finalization path: retain and emit the finding and its reproducer, then exit before corpus/run accounting. The cancelled `afterInvariant` and generic campaign-stop branches need the same treatment.

File: crates/evm/evm/src/executors/invariant/mod.rs [hunk 9]
```diff
@@ -1211,7 +1267,8 @@ impl<'a, FEN: FoundryEvmNetwork> InvariantExecutor<'a, FEN> {
                             InvariantFuzzError::MaxAssumeRejects(config.max_assume_rejects),
                         );
                         campaign_state.request_terminal_stop();
-                        break 'stop;
+                        stop_after_run = true;
```
Comment (grandizzy) [inline crates/evm/evm/src/executors/invariant/mod.rs]: **[P1] Do not count max-assume rejection as a completed run**

Breaking only the inner loop reaches `end_run` and increments `runs`, although the configured sequence and final invariant evaluation never completed. This changes the established result from `runs: 0` to `runs: 1` and can persist the rejected prefix; it is the behavior currently failing `should_not_fail_replay_assume`.

Restore the outer-loop exit here and preserve discard-selector metrics independently rather than accepting the run.

File: crates/evm/evm/src/executors/invariant/mod.rs [hunk 10]
```diff
@@ -1162,16 +1221,21 @@ impl<'a, FEN: FoundryEvmNetwork> InvariantExecutor<'a, FEN> {
...
-                    invariant_test.record_metrics(
+                if config.show_metrics
+                    && let Some(metric) = InvariantTest::call_metric(
                         current_run.inputs.last().expect("checked above"),
                         call_result.reverted,
                         discarded,
-                    );
+                    )
+                {
+                    current_run.metrics.push(metric);
```
Comment (grandizzy) [inline crates/evm/evm/src/executors/invariant/mod.rs]: **[P2] Retain metrics for calls completed before timeout**

Selector metrics are staged until the entire run is accepted, so timeout during a later call or predicate discards metrics for earlier handler calls that completed normally. This is why `should_apply_configured_timeout` now loses its `TimeoutHandler.increment` row.

Record selector metrics after confirming that the individual handler execution was not cancelled; keep run-owned corpus, coverage, and completed-run counters staged.

**Review by grandizzy — CHANGES_REQUESTED** [review]
The campaign cancellation signal is wired correctly now, but partial-run finalization still either drops completed findings/metrics or accepts interrupted runs as complete. The two failing invariant CLI tests expose these accounting regressions.

File: crates/evm/evm/src/executors/invariant/mod.rs (hunk 7 above)
Comment (stevencartavia) [reply crates/evm/evm/src/executors/invariant/mod.rs]: [4e7c84e](https://github.com/foundry-rs/foundry/pull/15689/commits/4e7c84e4d18b4f7d488fe3260777a5e3fb78bf41)

File: crates/evm/evm/src/executors/invariant/mod.rs (hunk 8 above)
Comment (stevencartavia) [reply crates/evm/evm/src/executors/invariant/mod.rs]: [4e7c84e](https://github.com/foundry-rs/foundry/pull/15689/commits/4e7c84e4d18b4f7d488fe3260777a5e3fb78bf41)

File: crates/evm/evm/src/executors/invariant/mod.rs (hunk 9 above)
Comment (stevencartavia) [reply crates/evm/evm/src/executors/invariant/mod.rs]: [4e7c84e](https://github.com/foundry-rs/foundry/pull/15689/commits/4e7c84e4d18b4f7d488fe3260777a5e3fb78bf41)

File: crates/evm/evm/src/executors/invariant/mod.rs (hunk 10 above)
Comment (stevencartavia) [reply crates/evm/evm/src/executors/invariant/mod.rs]: [4e7c84e](https://github.com/foundry-rs/foundry/pull/15689/commits/4e7c84e4d18b4f7d488fe3260777a5e3fb78bf41)

**Review by grandizzy — APPROVED** [review]
lgtm, ty!

**Review by mablr — APPROVED** [review]
